Potential fix for enums as class constants

jdwx opened the pull request:
Potential fix for #4870 

Not sure if this is the right place in the code or even the right way to approach this, but I added a test and it appears to work without breaking anything.

jdwx committed:
Merge pull request #1 from phan/v5

Update 2024-07-30

Commit:
Fix for incorrect warning on enum class constants.

Commit:
Fix a couple of trivial comment typos.

...I was in the neighborhood.

Commit:
Fix for generics.

Commit:
Handle enums in union types.

jdwx commented:
I've changed this to move the "is it an enum?" code to a separate function. This probably isn't the right place for it to live, but it should be easy to move now.

This allows the patch to work for union types containing an enum.

Commit:
Add union type to const enum comment test case.

TysonAndre commented:
I forgot Phan even had this issue type.

-----

https://wiki.php.net/rfc/new_in_initializers was also merged in PHP 8.1 (same PHP version that introduced enums) - you can have arbitrary objects in constants (e.g. `const X = new stdClass()`) as a result, not just enums.

The issue message should probably be changed to be a compatibility warning 

And this should skip emitting the issue if `Config::get_closest_minimum_target_php_version_id() < 80100` to avoid emitting it for projects that don't need to support PHP versions that are older than PHP 8.1.

jdwx commented:
I think the new_in_initializers change applies to global constants, not class constants. Whereas enums are explicitly allowed for class constants.

I.e.:

```php
<?php

declare( strict_types = 1 );

class Foo {
	public string $st = "bar";
}

const FOO = new Foo();

class Baz {
	public const Foo FOO = new Foo();
}
```

Gives:

```
PHP Fatal error:  New expressions are not supported in this context in /private/tmp/phan/test.php on line 12
```

I.e., the expression on line 9 is valid, but not the one on line 12. 

By contrast:

```php
<?php

declare( strict_types = 1 );

enum Foo {
	case bar;
}

const BAR = Foo::bar;

class Baz {
	public const Foo BAR = Foo::bar;
}
```

Works correctly in both cases.  So, at least for now, they might need to be addressed separately.

I'll see if I can add the version check to suppress my change for <8.1 to this pull request; I'll have to think about it for at least a few minutes to make sure I get that right. 

Also, just want to say it's great to see a note from you. Your hard work means a lot to a lot of people you'll never meet, even if they aren't really in a position to do much more about it than tell you that from time to time.

Commit:
Don't check for enums on PHP <8.1.

Tweaked the object comparison and added a couple of test cases
to make sure that an enum doesn't accidentally shadow a warning
from an object that has already been checked.

jdwx commented:
Happily, because there are no versions of PHP that support enums but not enums as class constants, adding the version check wasn't as complex as I feared. 

I also tweaked the subtype check and the test cases to make sure that an enum in a union type wouldn't accidentally hide a warning about an actual object in that type.

dasl- commented on src/Phan/Analysis/ClassConstantTypesAnalyzer.php:
```diff
@@ -105,4 +109,24 @@ public static function analyzeClassConstantTypes(CodeBase $code_base, Clazz $cla
     }
+
+
+    private static function typeIsEnum( CodeBase $code_base, Type $type ) : bool {
```
I noticed there is already: https://github.com/phan/phan/blob/b52a8829c841b31e2f03ff8f6768fc116c6e2746/src/Phan/Analysis/PostOrderAnalysisVisitor.php#L3722

It looks like `isTypeEnum` handles one edge case that your implementation does not. What do you think of moving the method to `Type::isEnum`. There is already a pattern of some methods in `Type` having a `CodeBase` argument, e.g.  https://github.com/phan/phan/blob/b52a8829c841b31e2f03ff8f6768fc116c6e2746/src/Phan/Language/Type.php#L2158

Probably we would have `Type::isEnum` return `false`, and then override the method in `ObjectType::isEnum`.

jdwx replied:
That makes a lot of sense. I had a similar thought about Type::isEnum() and ObjectType::isEnum(). But I don't have a lot of PR history with this project nor do I understand its innards well, so I figured it was better to start small and localized and then let things evolve to a more general solution if people who know more than me think that's best for the project.

jdwx replied:
@dasl- Let me ask a silly question here. Were you suggesting that I should do that and resubmit this pull request? Happy to do that if so; sometimes I'm a little slow on the uptake!

dasl- reviewed (commented):
>Also, just want to say it's great to see a note from you. Your hard work means a lot to a lot of people you'll never meet, even if they aren't really in a position to do much more about it than tell you that from time to time.

Well said! Curious what Tyson thinks of my comments also.

jdwx committed:
Merge branch 'phan:v5' into v5

rlerdorf commented:
I'm going to close this since it is on the v6 branch now. I don't think it is critical for v5. Thanks for your work on this.
